zebra-script/src/lib.rs

```diff
@@ -202,33 +149,30 @@ impl CachedFfiTransaction {
             input_index: n_in,
             sighasher: SigHasher::new(&self.transaction, branch_id, &self.all_previous_outputs),
         });
-        // SAFETY: The `script_*` fields are created from a valid Rust `slice`.
-        let ret = unsafe {
-            zcash_script::zcash_script_verify_callback(
-                (&*ctx as *const SigHashContext) as *const c_void,
-                Some(sighash),
-                lock_time,
-                is_final,
-                script_pub_key.as_ptr(),
-                script_pub_key.len() as u32,
-                signature_script.as_ptr(),
-                signature_script.len() as u32,
-                flags,
-                &mut err,
-            )
-        };
+        let ret = zcash_script::Cxx::verify_callback(
+            &|script_code, hash_type| {
+                let script_code_vec = script_code.to_vec();
+                Some(
+                    (*ctx).sighasher.sighash(
+                        HashType::from_bits_truncate(hash_type.bits() as u32),
+                        Some(((*ctx).input_index, script_code_vec)),
+                    ).0
+                )
+            },
```

Comment:
I currently get some lifetime errors around this callback. They make sense, but I don’t know how to avoid them.
```
error: lifetime may not live long enough
   --> zebra-script/src/lib.rs:153:13
    |
111 |       pub fn is_valid(&self, branch_id: ConsensusBranchId, input_index: usize) -> Result<(), Error> {
    |                       - let's call the lifetime of this reference `'1`
...
153 | /             &|script_code, hash_type| {
154 | |                 let script_code_vec = script_code.to_vec();
155 | |                 Some(
156 | |                     (*ctx).sighasher.sighash(
...   |
160 | |                 )
161 | |             },
    | |_____________^ cast requires that `'1` must outlive `'static`

error[E0597]: `ctx.input_index` does not live long enough
   --> zebra-script/src/lib.rs:158:31
    |
153 |               &|script_code, hash_type| {
    |               ------------------------- value captured here
    |  _____________|
    | |
154 | |                 let script_code_vec = script_code.to_vec();
155 | |                 Some(
156 | |                     (*ctx).sighasher.sighash(
157 | |                         HashType::from_bits_truncate(hash_type.bits() as u32),
158 | |                         Some(((*ctx).input_index, script_code_vec)),
    | |                               ^^^^^^^^^^^^^^^^^^ borrowed value does not live long enough
159 | |                     ).0
160 | |                 )
161 | |             },
    | |_____________- cast requires that `ctx.input_index` is borrowed for `'static`
...
170 |       }
    |       - `ctx.input_index` dropped here while still borrowed
```

Reply:
`verify_callback` constrains the closure with `Fn`. I feel like I should be able to weaken the constraint to `FnOnce`, but I end up with different errors (on the zcash_script side, rather than the zebra side) for both `FnOnce` and `FnMut`.

Reply:
It might work with `move |script_code, hash_type| {`, if `verify_callback()` can take an owned closure, I don't think making it a `FnOnce` or `FnMut` will help here since it's not consuming or mutating anything as far as I can tell, it looks like `ctx` just might be dropped early if it's not moved and if the closure is returned from this function or passed to another thread.

Reply:
Thanks for your assistance with this!